server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java

```diff
@@ -628,12 +628,16 @@ protected void parseCreateField(ParseContext context) throws IOException {
     }
 
     @Override
-    public String parseSourceValue(Object value) {
+    public String parseSourceValue(Object value, String format) {
         String date = value.toString();
         long timestamp = fieldType().parse(date);
-
         ZonedDateTime dateTime = fieldType().resolution().toInstant(timestamp).atZone(ZoneOffset.UTC);
-        return fieldType().dateTimeFormatter().format(dateTime);
+
+        DateFormatter dateTimeFormatter = fieldType().dateTimeFormatter();
+        if (format != null) {
+            dateTimeFormatter = DateFormatter.forPattern(format).withLocale(dateTimeFormatter.locale());
```

Comment:
I'd prefer to avoid this one day, but for now it is cool. And, I think it'd be a bit overboard to try and avoid it now. If I did want to avoid it maybe something like:
```
public Function<Object, String> sourceParser(String format);
```

I don't know. Just thinking out loud. Not needed now.

Reply:
I agree, instead of a function we could even introduce an object like `SourceValueParser`. This could also be nice because it'd help clean up some of the fragility around the dual `lookupValues` and `parseSourceValue` methods. It did feel like overkill to do this refactor now, though.